### DISCUSSION TITLE
custom links for interfaces in device view

### DISCUSSION THREAD

**Discussion by jcralbino** [netbox-community/netbox#6679] [Q&A] [answered]
Hello

I am exploring the custom links options and i have been able to make the links work when i reference a direct object. But i have a use case where i am interested in using the custom links where several objects are seen and i wanted to know if this is possible.

- In https://netboxhost/dcim/devices/201/interfaces/ , we have a list of interfaces. For some i want to create a custom link. The custom link will allow me to go directly to the Orchestrator(Cisco APIC in this case) and check status of that object. 
- In https://netboxhost/admin/extras/customlink/ i created the following custom link
  * Name: ACI_VPC_Device
  * Group name: ACI Links
  * Link text ```{% if obj.interfaces.filter(type='lag').get().type == 'lag' %}View ACI VPC{% endif %}```
  * Link url ```https://apichost/#c:d|uni/infra/funcprof/accbundle-{{ obj.interfaces.filter(type='lag').get() }}```

- this works well when only one lag is configured for this device. If i have more than one i get this:
![image](https://user-images.githubusercontent.com/44149262/123910291-8412e380-d97a-11eb-9dd8-5eaa4f14c570.png)

Will it be possible to create one link for each lag interface? 

I am still a newbie in the jinja2 templating so i am sure that my code can be improved, but i am not sure if netbox is ready to provide a different object for each interface that can be used in the link URL section.

**Comment by candlerb** [comment, marked as answer]
A custom link only generates one button, and it can only have one URL target.

However, you could create a custom link on `DCIM > interface` instead of on `DCIM > device`.  Starting at https://netboxhost/dcim/devices/201/interfaces/, the user could filter this down to LAGs(*), click on the interface of interest which would take them to (say) https://netboxhost/dcim/interfaces/1234/, and the custom link button would be there.

(*) In v2.11.7, filtering the list of interfaces is only provided by name, not by type or other attributes.  However if the interfaces have names like "lag 1", "lag 2" etc then typing "lag" in the filter box will work.

-----

The error occurs during your 'link text' expansion:

```
Link text: {% if obj.interfaces.filter(type='lag').get().type == 'lag' %}View ACI VPC{% endif %}
```

`obj.interfaces.filter(type='lag')` returns multiple items, and `get()` only works on a single item.  You can make the expansion work with something this (untested):

```
Link text: {% if obj.interfaces.filter(type='lag').count() > 0 %}View ACI VPC{% endif %}
```

... but you'd still have the same problem when expanding the URL.  You could use `first()` to get one, but I don't think that's what you want.

**Reply by jcralbino** [reply]
Indeed i used the same strategy for the https://netboxhost/dcim/interfaces/1234/ and there it works well. We have a link provided in case the interface is a lag.

Thanks for the tip for 2.11.7 i will change to use a name, in this case it starts always with vpc. 

My issue is still on expanding the url.
In the URL field i cannot run full jinja2 expressions, to have a for looping the results and create a url for each one of them?

**Reply by candlerb** [reply]
As I said already, you can't do that.  One "custom link" = one button = one URL only.

**Reply by jcralbino** [reply]
That was clear to me. 
My remark was aimed to understand if this could be possible in the future, if required i can open a feature request. But indeed we can always create the links where the result is aimed to only one object.

**Reply by jcralbino** [reply]
I have ended up with this code:
 * Link test: 
 ``` 
 {% if obj.interfaces.filter(name__icontains='vPC').count() == 1 %}
  View VPC in ACI
 {% elif obj.interfaces.filter(name__icontains='vPC').count() > 1 %}
  View first VPC in ACI
 {% endif %}
```
* Link URL:
```
https://apichost/#c:d|uni/infra/funcprof/accbundle-{{ obj.interfaces.filter(name__icontains='vPC').first() }}
```

**Reply by jcralbino** [reply]
Thanks @candlerb  for the feedback. Really appreciated.

**Comment by jcralbino** [comment]
It seems i cannot use this validation on the custom link that is working in the nbshell()
```
>>> type(obj)
<class 'dcim.models.devices.Device'>
>>> type(obj.interfaces.filter(name__icontains='vPC').first().untagged_vlan)
<class 'NoneType'>
>>> (obj.interfaces.filter(name__icontains='vPC').first().untagged_vlan) is not None
False
```

In the link Text i have:
```
{% if (
       (obj.interfaces.filter(name__icontains='vPC').first().untagged_vlan is not None ) and
       (obj.interfaces.filter(name__icontains='vPC').first().untagged_vlan.name.find('BD') == 0)
       )
%}
 View {{ obj.interfaces.filter(name__icontains='vPC').first().untagged_vlan.name }} in ACI
{% endif %}
```

I get this error:
![image](https://user-images.githubusercontent.com/44149262/123982446-54d49480-d9c3-11eb-9fe9-7ac298fce4af.png)

**Reply by jcralbino** [reply]
@candlerb  is this a bug? Shall i open a bug report or is it expected?

**Reply by candlerb** [reply]
It's expected.  [Jinja2](https://jinja.palletsprojects.com/en/3.0.x/templates/) is not Python; the condition "is not None" is not valid Jinja2, as the error says.

However there is a test `none()`, and there is `is sameas none`.  See docs [here](https://jinja.palletsprojects.com/en/3.0.x/templates/#tests).  And see this note:

> Note
> The special constants true, false, and none are indeed lowercase. Because that caused confusion in the past, (True used to expand to an undefined variable that was considered false), all three can now also be written in title case (True, False, and None). However, for consistency, (all Jinja identifiers are lowercase) you should use the lowercase versions.

Google for "jinja2 live parser" to find some [websites](https://cryptic-cliffs-32040.herokuapp.com/) where you can test your Jinja2 expressions.

**Reply by jcralbino** [reply]
Thanks for the clarification. I will test the jinja2 templates using your feedback

**Comment by jcralbino** [comment]
Is there anyway that i can place the link in the middle?
It seems that if the text is long enough it will not be rendered correctly. 
In this case i am doing the following attempt:

* Link text:
```
{% if obj.interfaces.filter(name__icontains='vPC').count() == 1 %}
 View VPC in ACI
{% elif obj.interfaces.filter(name__icontains='vPC').count() > 1 %}
 View {{ obj.interfaces.filter(name__icontains='vPC').first().name }} in ACI
{% endif %}
```

The result is this where the text is now shown completelly:
![image](https://user-images.githubusercontent.com/44149262/123918416-25eafe00-d984-11eb-9238-c1dfcd856eba.png)

**Reply by jcralbino** [reply]
@candlerb regarding this what is the maximum size that we can use ? Or can we control where  the group link is placed

**Reply by candlerb** [reply]
I'm unaware of any maximum size in terms of characters, and the font is proportionally-spaced anyway.

Your screenshots show that very long strings don't render well.  If you can find a good CSS solution to fix that, then by all means provide a patch.
